drafts/draft-0005.md

```diff
@@ -0,0 +1,188 @@
+
+Additionally, there may be use cases where users do not want their multisig transactions broadcasted to a public coordinator, but would still like to use existing wallets to submit them to a private coordinator.
+
+## Specification
+
+### Account Data
+
+A multisig account adds a data entry with the key `multisig.domain` and the value of a hostname used to resolve the `stellar.toml` file of the multisig service.
+
+Example:
+Key: `multisig.domain`
+Value: `stellarguard.me`
+
+## Multisig Server stellar.toml
+
+Multisig coordinators advertise the existence of their service through the `stellar.toml` file. The top-level parameter `MULTISIG_ENDPOINT` should contain a fully-qualified URL of a multisig coordination service where transactions should be submitted.
+
+Example of `stellar.toml`:
+```
+MULTISIG_ENDPOINT="https://stellarguard.me/api/transactions"
+```
+
+### Multisig Server API Endpoints
+
+Multisig Services must expose two REST API endpoints.
+
+#### Submit Transaction Endpoint
+
+This is the endpoint defined in the `stellar.toml` under the `MULTISIG_ENDPOINT` key.
+
+##### Request
+
+### Request
+
+```
+POST MULTISIG_ENDPOINT
+```
+
+ContentType: `application/json`.
+
+Request Body Parameters:
+
+Name | Type | Description
+-----|------|------------
+`uri` | string | A [SEP-0007 style](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0007.md)  `tx` URI for the transaction to sign.
+
+Example:
+
+`POST https://stellarguard.me/api/transactions `
+
+```json
+{
+  "uri": "web+stellar:tx?xdr=AAAAAHFd0%2BHQV5u%2FY%2FfM3%2BVelUr1IWwSFL8CUDIAUudUdD4MAAAAZAADyI8AAAADAAAAAAAAAAAAAAABAAAAAAAAAAEAAAAAcV3T4dBXm79j98zf5V6VSvUhbBIUvwJQMgBS51R0PgwAAAAAAAAAAACYloAAAAAAAAAAAVR0PgwAAABAEw8ODG0iixkbAHg1aJATAnZS2531PhGauuSvFDad2WxHKzIenUNbc7K5mGiSpe5jvqe19OQCbNFuBjqN11jfBw%3D%3D"
+}
+```
+
+##### Response
+
+On success, the endpoint should return a 200 HTTP status code and a JSON-encoded object.
+
+Response fields:
+
+Name | Type | Description
+-----|------|------------
+`id` | string | Required. An identifier for the multisig request that is generated by the multisig service.
+`statusHref` | string | Required. A fully qualified url where an agent can poll for updates on the status of the multisig request. This is specified in the Status Endpoint section.
+`extras` | object | Optional. An object that contains additional application specific response data which could be used to provide an enhanced user experience. For instance, showing a link to authorize a transaction.
+
+Example:
+
+```json
+{
+  "id": "c585ceee-b1b9-4009-aa37-b8544346a036",
+  "statusHref": "https://stellarguard.me/api/transactions/c585ceee-b1b9-4009-aa37-b8544346a036",
+  "extras": {
+    "isStellarGuard": true,
+    "url": "https://stellarguard.me/transactions/c585ceee-b1b9-4009-aa37-b8544346a036"
+  }
+}
+```
+
```

Comment:
This seems too much complicated to me: why introducing a new kind of id when we already have txHashes ?

Reply:
> You can have two different signing requests that have the same tx hash but the callback is different

That's weird to introduce complexity for such a far-fetched edge case. If providing multiple callbacks is really important (which I doubt) it would be better to allow to do so into the POST request. It's the same transaction and you probably don't want to collect signatures twices / have two entries for it in your system.

> think of an enterprise that is using Stellar that has their own internal multisig service that does not need to interoperate

Isn't interoperability the point of writing a protocol ? Why an enterprise needs to be told how to implement its own internal signature collector?

Reply:
I think this is not resolved yet. Being able to register several time the same transaction comes with a lot of trickyness.

The main one is that it would make coordination complicated in case each co-signer received the tx request independently: because instead of coordinating the signatures under the same entry (by txHash) it will create as much entries as signatures are sent (by id).

It raises a lot of question: must the signatures shared between different entries with the same txHash? What is the cost of implementing the required checks? Is it OK to use signature of someone who signed tx with callback1 for validating tx with callback2? Can't I abuse this system by sending the same tx request again and again & change only a few characters?

Being able to have the same tx registered several times will also complicate everybody's work; because you'll have to handle that special case at every level (coordinator, wallets, utils).

Some weeks ago I pointed to @pselden  that if I sent the same tx request twice at StellarGuard it would create two different entries for that transaction. Then if I validate one the other is still there. It shows that this kind of stuff is not so easy to deal with: there's a lot of logic to get right so the user gets what he'd expect.

In short: this is a lot of troubles for a feature that will probably not really get used (multiple callbacks).

Now if you go with identification by txHash it's all solved from start. So I maintain this is the way to go.

Reply:
Indeed my point was not to discuss StellarGuard design but to exemplify what will happen if coordinator use arbitrary IDs. StellarGuard accounts only have one external co-signer, now what do you think will happen if 10 co-signers are posting signatures for a given transaction?

You're saying you don't want to enforce coordinators to use txHash: why wouldn't you want to do that if it actually solves a whole group of issues and makes simpler?

Some of the issues that come to mind:

* Duplicates entries for the same transaction
* Impossible to probe endpoint for a given transaction status prior to know its ID
* Force wallet to track additional data (more data to encrypt/store/submit)
* Make changing coordinator a complicated process (it's going to be a complete mess wallet-side)

Advantages of using txHash:

* Predictability of entry ID: once a wallet receive an external transaction request it can immediately probe its status
* Changing coordinator doesn't affect TxPool state
* Signatures are properly aggregated under the same entry

Reply:
These are good arguments. Especially compelling is the fact that that you won't have to store any other information about the transaction besides the hash.

Some potential downsides to weigh:
- How do you handle submitting the same resubmitting the transaction with other parameters (callback?).
- How do you handle resubmission of transactions in the even the first one failed and or was rejected by signers?

These could be left up to the coordinator without much impact though (they could decide whether they accept it, overrwrite it, allow a duplicate after the first one is rejected).

That said, @nikhilsaraf I'm convinced that there's enough benefits to make this change. Your thoughts?

Reply:
the points brought up for using tx hash for **coordination** makes sense, however I want to stress again that multisig coordination is a _specfic case of multisig accounts_ and this bootstrapping SEP is trying to accommodate all cases.

When the multisig coordination SEP comes out, it will be a good avenue to discuss whether coordinators should use tx hashes or not. This is the reason why this SEP supports a **string** ID type which is more general and flexible.

As for storing the ID, that's a convenience that the wallet can (and probably should) provide to it's users. Are they not already storing some information related to the user's account? If they don't want to store anything then they can always additionally allow the user to enter an ID that is retrieved from the vendor's website.

I think we've already gotten way into the weeds here with this one, a string ID is very flexible and if that doesn't work out we can always amend this SEP accordingly.

@pselden added my comments here from our discussion. My understanding is that you support using string IDs as well, please feel free to unresolve this thread and add a comment if I misunderstood that from our discussion.

Reply:
Yes, I support string IDs. The tipping point for me was your point that this doesn't _have_ to be use for coordination at all, and reiterating that this is not about the coordinator.

Reply:
I sincerely don't understand your point and I don't think this is resolved. Getting into the details of how this is going to work is required to take the good decisions. The problems I pointed to have not been addressed so far - thus it is not correct to set this point as resolved.

Reply:
I sincerely don't understand your point and I don't think this is resolved. Getting into the details of how this is going to work is required to take the good decisions. The problems I pointed to have not been addressed so far - thus it is not correct to set this point as resolved.